Feature/#204 스터디 생성 모달 구현

yeonddori opened the pull request:
### 관련 이슈

- [Feature/#205 DatePicker 구현](https://github.com/BDD-CLUB/01-doo-re-front/pull/209) 선행 리뷰 부탁드립니다!
- close #204

### 작업 요약

- 스터디 생성 모달을 구현했습니다.
- curriculums를 어떻게 할지 결정됨에 따라 다시 수정하겠습니다!

### 작업 상세 설명

- 땅님의 구현처럼 필수 입력 란은 포커스 해제되었을 시 알림이 뜨도록 하였습니다! 이때 Selector의 onBlur 속성을 추가하게 되었고, 비동기화 업데이트 때문에 `useRef`를 사용하게 되었습니다.
- 필수 입력 란이 입력이 되지 않았을 시 다음 버튼으로 넘어가지 않도록 하였습니다. 

### 리뷰 요구 사항

- 모달이 총 3단계가 있기에, 모두 크기가 다르면 이상할 것 같아서 일단 `<Box overflowY="auto" minH="60vh" maxH="60vh">`를 추가하였는데, 크기는 어떻게 하는 것이 좋을지 조율하면 좋을 것 같습니다.
- 그리고 폴더 위치도 조금 헷갈립니다..! 스터디 생성은 팀페이지에서 했던 것 같은데, 그럼 `team` 폴더로 옮겨야 할까요?

```tsx
'use client';

import { Box } from '@chakra-ui/react';
import { useState } from 'react';

import CreateStudyModal from '@/containers/study/CreateStudyModal';

const App = () => {
  const [isModalOpen, setIsModalOpen] = useState(true);

  return (
    <Box w="300px" p={20}>
      <CreateStudyModal isOpen={isModalOpen} setIsModalOpen={setIsModalOpen} />
    </Box>
  );
};

export default App;
```

- 리뷰 예상 시간 : 15분

### 미리 보기

![스크린샷 2024-05-02 171013](https://github.com/BDD-CLUB/01-doo-re-front/assets/126975394/13eb8385-5943-446c-8d7c-dca884ae186e) ![스크린샷 2024-05-02 171023](https://github.com/BDD-CLUB/01-doo-re-front/assets/126975394/8edc4263-f1ed-40fd-8a40-5041d44bc807) ![스크린샷 2024-05-02 171032](https://github.com/BDD-CLUB/01-doo-re-front/assets/126975394/0aff7d1d-a4d7-4e5b-8918-d337b99880c6)

![2024-05-02165535-ezgif com-video-to-gif-converter](https://github.com/BDD-CLUB/01-doo-re-front/assets/126975394/d30af1b5-c5a1-4cb8-9423-28ac88677ada)

yeonddori committed:
feat: onBlur 속성 추가

#204

yeonddori committed:
feat: 스터디 생성 모달 구현

#204

yeonddori committed:
feat: api에 맞게 데이터 필드 추가

#204

yeonddori committed:
feat: import CurriculumItemDto

#204

jasper200207 reviewed (approved):
👍

> curriculums을 map으로 돌릴 때, key 값을 index로 할 생각이었으나 권장하지 않는 방법이라고 하기에 일단 curriculums에 id 속성도 추가하였습니다. api 넘길 때는 따로 curriculums.name만 모은 리스트로 넘겨줘야 할 것 같은데, 혹시 이보다 좋은 방법이 있을까요?

제가 알기로는 index로 하면 안되는 이유가 react에서 key는 리렌더링 대상을 정할때 사용하는데 index의 값은 리스트의 요소가 바뀌면 바뀔수 있기 때문으로 알고있습니다.
예를 들어 [ 1, 2 , 3, 4 ] 에서 2를 삭제한다고 했을 때 index를 기준으로 하면 3, 4의 인덱스가 모두 바뀌어 리렌더링이 필요하지만 값을 기준으로 하면 2만 리렌더링으로 사라지고 3, 4는 영향이 없는걸로 알고 있습니다. 
그래서 가장 간단한 해결방법은 그냥 리스트 안의 값을 넣는 것입니다. 이경우는 별다른 추가요소 없이 쉽게 할 수 있지만, 커리큘럼 이름이 같으면 key값 중복이 되는 문제가 발생합니다. 
다른 해결방법은 react 공식 문서에도 알려주는 방식인데 uuid를 페이지에서 생성해서 사용하는 방식입니다. 주의할 점은 렌더링과정에서 생성하면 안됩니다. ( key={Math.random} ) 즉, id 값을 커리큘럼이 추가될때마다 임의로 만들어서 사용해야할 듯 합니다. 예시 코드는 아래에 있습니다.

```typescript

const handleAddCurriculum = () => {
    setCurriculumItems([
      ...curriculumItems,
      {
        id: uuid(),
        name: '',
        itemOrder: 0,
        isDeleted: false,
      },
    ]);
  };

return (
...
{curriculumItems.map((curriculum, index) => (
                <Input
                  key={curriculum.id}
                  textStyle="bold_xl"
                  _placeholder={{ color: 'white' }}
                  onChange={(e) => handleCurriculumChange(index, e.target.value)}
                  placeholder="커리큘럼명"
                  value={curriculum.name}
                />
))}
...
)

```
참고로 uuid 모듈은 npm i uuid로 설치 가능합니다.

> 그리고 폴더 위치도 조금 헷갈립니다..! 스터디 생성은 팀페이지에서 했던 것 같은데, 그럼 team 폴더로 옮겨야 할까요?

아마 폴더 구조의 기준이 사용하는 페이지 였던걸로 기억합니다! 팀페이지에서 사용하면 team 폴더가 맞을듯 합니다.

yeonddori commented:
인덱스를 쓰지 않는 이유는 처음 알았네요. 감사합니다!!
커리큘럼은 따로 분리하자는 말도 나와서, 결정된다면 다시 수정하겠습니다!

yeonddori committed:
fix: 커리큘럼 추가 제거

#204

yeonddori committed:
Merge branch 'develop' of https://github.com/BDD-CLUB/01-doo-re-front into Feature/#204-스터디_생성_모달_구현

yeonddori committed:
refactor: 폴더 위치 변경 study -> team

#204

jasper200207 reviewed (approved):
👍

llddang commented on src/components/Selector/index.tsx:
```diff
@@ -17,7 +17,7 @@ const Selector = ({ selected, label, handleSelector }: SelectorProps) => {
 
   return (
-    <Menu>
+    <Menu onClose={onBlur}>
```
여기서 close 용도로 사용되는데 변수명이 'onBlur'인게 어색한 것 같습니다.
혹시 handleClose와 같은 변수명은 어떻게 생각하시나요?

llddang commented on src/containers/team/CreateStudyModal/index.tsx:
```diff
@@ -0,0 +1,145 @@
+};
+
+const CreateStudyModal = ({ isOpen, setIsModalOpen }: CreateStudyModalProps) => {
+  const [step, setStep] = useState(1);
```
여기서도 type <number> 지정하는 것이 좋아보입니다!

llddang commented on src/containers/team/CreateStudyModal/index.tsx:
```diff
@@ -0,0 +1,145 @@
+import { useEffect, useRef, useState } from 'react';
+
+import AutoResizeTextarea from '@/components/AutoResizeTextarea';
+import StyledDatePicker from '@/components/DatePicker';
```
혹시 여기서 에러뜨지 않나요?
브랜치에서 확인해보니 해당 경로가가 존재하지 않습니다..!

pipisebastian replied:
[DatePicker구현](https://github.com/BDD-CLUB/01-doo-re-front/actions/runs/9111645718/job/25049198244)한게 디벨롭에 안올려져있더라구요..! 
`npm install --save @types/react-datepicker` 로 typescript버전도 같이 설치해줘야할 것 같습니다!

jasper200207 replied:
> [DatePicker구현](https://github.com/BDD-CLUB/01-doo-re-front/actions/runs/9111645718/job/25049198244)한게 디벨롭에 안올려져있더라구요..! `npm install --save @types/react-datepicker` 로 typescript버전도 같이 설치해줘야할 것 같습니다!

react-datepicker type이 없어 build에서 오류가 났습니다! (action이 안돌았어요...)
위의 내용과 같이 `@types/react-datepicker` 가 필요합니다. typescript 관련은 devDependencies 로 넣어도 괜찮아서
`npm i --save-dev @types/react-datepicker`로 설치해도 될듯합니다

yeonddori replied:
> 혹시 여기서 에러뜨지 않나요? 브랜치에서 확인해보니 해당 경로가가 존재하지 않습니다..!

pr을 먼저 올리고 나중에 datepicker가 merge돼서 그렇습니다..!! pull 받겠습니다!
이제 보니 제가 뒤에 이름을 수정했어서 `@/components/StyledDatePicker`가 되어야 하는데 `@/components/DatePicker`로 되어있었네요..!

> [DatePicker구현](https://github.com/BDD-CLUB/01-doo-re-front/actions/runs/9111645718/job/25049198244)한게 디벨롭에 안올려져있더라구요..! `npm install --save @types/react-datepicker` 로 typescript버전도 같이 설치해줘야할 것 같습니다!

저도 처음엔 설치했다가 `@types` 붙은 게 다른 외부 모듈은 안쓰는 것 같아서 지웠었는데 다시 받겠습니다 ㅎㅎ 감사합니다!

llddang reviewed (approved):
수고하셨습니다!

yeonddori committed:
Merge branch 'develop' of https://github.com/BDD-CLUB/01-doo-re-front into Feature/#204-스터디_생성_모달_구현

yeonddori committed:
chore: @types/react-datepicker 추가

#204

yeonddori committed:
refactor: 변수명 변경 onBlur -> handleClose

#204

yeonddori committed:
refactor: useState type 지정

#204

yeonddori committed:
Merge branch 'develop' of https://github.com/BDD-CLUB/01-doo-re-front into Feature/#204-스터디_생성_모달_구현

pipisebastian commented on src/containers/team/CreateStudyModal/index.tsx:
```diff
@@ -0,0 +1,145 @@
+  const [alertDescription, setAlertDescription] = useState<boolean>(false);
+  const [alertSelectedCropId, setAlertSelectedCropId] = useState<boolean>(false);
+
+  const crops = ['토마토', '고구마', '당근', '완두콩', '벼'];
```
crops정보의 경우, 스터디 수정 모달이나, 작물 정보를 볼때도 불러와야하니까, 
필요한 파일마다 해당 코드를 복붙해와야해서 조금 번거로울 것 같아요!

그리고 '토마토'의 id를 뭐로할지는, 저번에 백엔드에서 프론트 맘대로 결정해주세요! 라고 했단말이


아예 constants파일을 만들어서 
```
const CROP = [
  { id: 1, name: '토마토' },
  { id: 2, name: '오이' },
  { id: 3, name: '당근' },
];
```
이처럼 id, name 객체를 만들어서 여러 파일에서 가져오는게 좋아보여용!

pipisebastian replied:
그렇게 되면 
selector에서 lable은 `label={CROP.map((c) => c.name)}` 로 하고,
handleSelector에서는  `setCropId(CROP.find((c) => c.name === value)?.id ?? 0);` 이렇게 수정하면 될 것 같습니다! 더 좋은방식 있으시면 말해주세용!

yeonddori replied:
피피님 말씀대로 따로 빼는 게 좋은 것 같아요!!
근데 constants 파일은 어디다 만드는 게 좋을까요..? `constants` 폴더에 `crop.ts`로 만들어야 할까요?
`constants` 폴더 내 파일이 color와 size라서 애매하네요..!

pipisebastian replied:
음음 말씀하신  `constants` 폴더에 `crop.ts` 괜찮아보입니당!
 color와 size가 ui관련이라 애매하긴 한데,, 일단 넣어놓고 나중에 ui관련(color, size) + 개발할때 쓰이는 상수(crop...) 들이 너무 많아지면 `constants`폴더에 또 폴더 나눠서해도 좋을지두용..! 더 좋은 방법은 생각안납니닷..

pipisebastian commented on src/containers/team/CreateStudyModal/index.tsx:
```diff
@@ -0,0 +1,145 @@
+  const [step, setStep] = useState<number>(1);
+  const [name, setName] = useState<string>('');
+  const [description, setDescription] = useState<string>('');
+  const [crop, setCrop] = useState<string>('작물을 선택해주세요');
```
해당 crop state는 사실 작물이름을 나타내는 거니까,
밑에 26 line에선 cropId로 지었으니까,이것과 비슷하게 cropName은 어떨까요?

그리고 '작물을 선택해주세요'는 보충설명의 텍스트라서 crop state에 기본값으로 있는게 조금 어색해보여요!
아예 빈문자열로 두고, 밑에 조건문(44line)에서도 빈문자열로 판단하는거 어떨까용?

pipisebastian replied:
그렇게되면, 현재 Selector에 기본으로 들어갈 placeholder를 추가해줘야할 것 같아요!
현재는 Selector의 값이 placeholder의 기능과, 진짜 값의 기능 2개다 하고있어서, 분리하는게 좋아보입니당..!!

pipisebastian replied:
```tsx
const Selector = ({ placeholder, selected, label, handleSelector, handleClose }: SelectorProps) => { // 수정
  const menuRef = useRef<HTMLDivElement>(null);
  const [menuWidth, setMenuWidth] = useState('0px');

  useEffect(() => {
    if (menuRef.current) {
      setMenuWidth(`${menuRef.current.offsetWidth}px`);
    }
  }, []);

  return (
    <Menu onClose={handleClose}>
      <MenuButton
        ref={menuRef}
        as={Button}
        w="100%"
        color="white"
        textAlign="left"
        bg="orange_light"
        _hover={{ bg: 'orange_light' }}
        _active={{ bg: 'orange_light' }}
        _focus={{ bg: 'orange_light' }}
        rightIcon={<BiChevronDown size="28px" />}
      >
        <Text textStyle="bold_md">{selected || placeholder}</Text>  // 수정
      </MenuButton>
      <MenuList
        overflow="hidden"
        minW={menuWidth}
        p="0"
        bg="orange_light"
        borderColor="orange_light"
        borderRadius="2xl"
      >
        {label.map((item, index) => (
          <MenuItem
            key={item}
            pl="15"
            color="white"
            bg="orange_light"
            _hover={{ bg: 'orange_dark' }}
            onClick={() => handleSelector(item)}
            value={index}
          >
            <Text textStyle="bold_md">{item}</Text>
          </MenuItem>
        ))}
      </MenuList>
    </Menu>
  );
};

```
placeholder를 받도록 확장해봤는데,
논리합 연산자(||)를 사용하여, selected가 Falsy값일때 placeholder를 넣고, 아닐때 selected를 반환을 해줬습니다.
일단 ?? 연산자가 아닌 ||를 사용한 이유는 selected가 기본적으로 빈 문자열('')(타입이 string)이고, 빈 문자열('')은 Falsy값으로 쳐주기때문에 사용했습니다!

pipisebastian replied:
타입의 경우도 이렇게 수정했습니다!
```tsx
export interface SelectorProps {
  placeholder?: string;
  selected: string;
  label: string[];
  handleSelector: (data: string) => void;
  handleClose?: () => void;
}

```

yeonddori replied:
> 해당 crop state는 사실 작물이름을 나타내는 거니까,
> 밑에 26 line에선 cropId로 지었으니까,이것과 비슷하게 cropName은 어떨까요?

이 부분은 api를 보고 뒤늦게 cropId를 추가했어서 저도 cropName이 더 좋은 것 같아요!! 지적 감사합니다!

> placeholder를 받도록 확장해봤는데, 논리합 연산자(||)를 사용하여, selected가 Falsy값일때 placeholder를 넣고, 아닐때 selected를 반환을 해줬습니다. 일단 ?? 연산자가 아닌 ||를 사용한 이유는 selected가 기본적으로 빈 문자열('')(타입이 string)이고, 빈 문자열('')은 Falsy값으로 쳐주기때문에 사용했습니다!

빈 문자열도 Falsy로 쳐주는 건 처음 알았어요!! 덕분에 배워갑니다~! 수정해놓을게요☺️

pipisebastian commented on src/containers/team/CreateStudyModal/index.tsx:
```diff
@@ -0,0 +1,145 @@
+  const [endDate, setEndDate] = useState<Date | null>(null);
+  const [alertName, setAlertName] = useState<boolean>(false);
+  const [alertDescription, setAlertDescription] = useState<boolean>(false);
+  const [alertSelectedCropId, setAlertSelectedCropId] = useState<boolean>(false);
```
이게 사실 이름, 설명, 작물 선택이 되었는지 안되었는지 boolean값인데, 이름이 alertName이라 주의 문구가 처음에 생각납니다!(ex. "이름을 입력해주세요!")
그래서 boolean을 더 잘나타낼 수 있는 `alertName` -> `isValidName` 은 어떨까요??

yeonddori replied:
이 부분은 땅님의 [Feature/#186 팀 생성 모달 구현](https://github.com/BDD-CLUB/01-doo-re-front/pull/199)에서 나온 의견을 수렴함에 따라, 똑같이 통일한 겁니다..! 바꾸면 다 같이 바꾸는 게 좋을 듯한데, 혹은 배포 이후 리팩토링 때 바꾸는 건 어떻게 생각하시나요..?!

pipisebastian replied:
아하 넹넹!! 좋씁니다!

pipisebastian reviewed (commented):
늦은 리뷰입니닷,,,!

pipisebastian commented:
>그리고 폴더 위치도 조금 헷갈립니다..! 스터디 생성은 팀페이지에서 했던 것 같은데, 그럼 team 폴더로 옮겨야 할까요?

스터디 폴더가 좋아보여요! 그리고 스터디 생성, 수정 모달은 파일을 각각 안만들고, 하나의 파일로도 충분히 가능할것같아요! 
조건에 따라 `flag ? 스터디 생성 : 스터디 생성` 이렇게 나눠준다면, 중복되는 코드가 더 줄어들 것 같습니다!!

flag의 경우, 만약 스터디 수정이라고 한다면, 스터디 수정에서는 모달창 props에 스터디 정보를 가지고 와야할테니까(조회 api)
스터디 정보가 있는지 여부로 생성모달인지, 수정모달인지 판별가능할 것 같습니당!

조금더 예시를 들자면, 이건 다른 프로젝트지만, 추가, 수정을 하나의 파일에서 하고 있어요!
![image](https://github.com/BDD-CLUB/01-doo-re-front/assets/81643702/50efcbfc-9239-48c0-96f3-35d301d38610)

yeonddori committed:
feat: Selector placeholder 속성 추가

#204

yeonddori committed:
feat: 시작날짜 필수값 지정 및 Selector placeholder에 따른 상태값 변경

- 작물 state 이름 변경 crop -> cropName
- Selector placeholder에 따라 cropName 초기값 빈 문자열로 변경
- 시작날짜 필수값 지정

#204

yeonddori commented:
> > 그리고 폴더 위치도 조금 헷갈립니다..! 스터디 생성은 팀페이지에서 했던 것 같은데, 그럼 team 폴더로 옮겨야 할까요?
> 
> 스터디 폴더가 좋아보여요! 그리고 스터디 생성, 수정 모달은 파일을 각각 안만들고, 하나의 파일로도 충분히 가능할것같아요! 조건에 따라 `flag ? 스터디 생성 : 스터디 생성` 이렇게 나눠준다면, 중복되는 코드가 더 줄어들 것 같습니다!!
> 
> flag의 경우, 만약 스터디 수정이라고 한다면, 스터디 수정에서는 모달창 props에 스터디 정보를 가지고 와야할테니까(조회 api) 스터디 정보가 있는지 여부로 생성모달인지, 수정모달인지 판별가능할 것 같습니당!
> 

딘님의 말씀처럼, 저희가 폴더 구조의 기준을 사용하는 페이지로 잡았던 걸로 기억해서 스터디에서 팀으로 옮기긴 했는데, 스터디 수정과 함께 사용한다면 애매해지긴 하네요..! 스터디 생성 모달과 스터디 수정 모달을 같이 쓰는 것은 저도 동의하는 바라 한 번 논의가 필요할 것 같습니다!

pipisebastian reviewed (approved):
수고하셨씁니당~~!

yeonddori committed:
refactor: 작물 정보 constants로 분리

#204

yeonddori committed:
refactor: 불필요한 코드 제거

#204
